### PR TITLE
LibraryCallMonitor: rewrite

### PR DESCRIPTION
**Pull request by adrianherrera**
The current `LibraryCallMonitor` plugin was broken if the imported DLL was not available in memory and had to be read from the guestfs. In this case, the addresses in the import address table are not valid runtime addresses and so hooking these addresses with the `FunctionMonitor` plugin was pointless (because these addresses would be overwritten at load time with the actual runtime address of the import).

The new `LibraryCallMonitor` plugin uses the `ModuleMap` plugin to keep track of loaded modules. When a library call occurs, we search the loaded modules to find if one of them contains the address that was called. If it does, we can find this function in the module's exports and determine the name of this function. Exports are cached for faster lookup.

### CODE REVIEW

File: src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.cpp [hunk 1]
```diff
@@ -1,145 +1,208 @@
...
-            exit(-1);
-        }
-        m_trackedModules.insert(*it);
+    static PluginState *factory(Plugin *plugin, S2EExecutionState *state) {
+        return new LibraryCallMonitorState();
     }
 
-    m_detector->onModuleLoad.connect(sigc::mem_fun(*this, &LibraryCallMonitor::onModuleLoad));
+    /// Add a new export to the cache at the given address
+    void put(uint64_t address, const Export &exp) {
```
Comment (vitalych) [inline src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.cpp]: I presume many plugins would like to have this info.
Would it make more sense to add this to ModuleMap instead? And then initialize this info lazily for interested clients?

File: src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.h [hunk 2]
```diff
@@ -1,82 +1,88 @@
...
+class ModuleMap;
 class OSMonitor;
+class ProcessExecutionDetector;
+class Vmi;
 
+///
+/// \brief Monitors external library function calls.
+///
+/// This plugin can be used to determine what external library calls a particular module makes at run time. The
+/// modules to monitor are those defined in the \code ProcessExecutionDetector plugin's config.
```
Comment (vitalych) [inline src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.h]: by modules you mean library or process main binaries?
Can it also track libraries imported by other libraries?

File: src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.h [hunk 3]
```diff
@@ -1,82 +1,88 @@
...
+/// The plugin works as follows:
+///   \li All indirect calls are monitored. Library calls are always made indirectly (via either the PLT in ELF
+///       binaries or the IAT in PE binaries)
+///   \li When an indirect call is made, we look up the module that contains the call target address
+///   \li Once we find that module, we can search its export table to find an entry that matches the call target
+///       address. If a match is found, report it
+///
+/// Note that this approach does not use the callee module's import table. This means that library calls that are not
+/// listed in the import table are still monitored. This is beneficial for malware analysis, where the import table
+/// is often destroyed and library calls are made "indirectly" (e.g. via \code LoadLibrary and \code GetProcAddress).
```
Comment (vitalych) [inline src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.h]: good!

File: src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.cpp (hunk 1 above)
Comment (adrianherrera) [inline src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.cpp]: Yep, this makes sense!

File: src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.h (hunk 2 above)
Comment (adrianherrera) [inline src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.h]: Hmmm good question. The answer is whatever the ProcessExecutionDetector can track. But I'll test and give you a real answer.

Comment (adrianherrera) [inline src/s2e/Plugins/ExecutionMonitors/LibraryCallMonitor.h]: I just did some tests by adding things like "kernel32.dll" and "crypt32.dll" to the moduleNames entry in the ProcessExecutionDetector config.

The answer is yes, this does allow you to track libraries imported by other libraries if they are specified in the ProcessExecutionDetector config.

**Comment by adrianherrera** [comment]
Ok, I moved the export cache to the ModuleMap plugin into a separate commit.

**Pull request merged**